modules/gapi/src/executor/gapi_tbb_executor.cpp

```diff
@@ -0,0 +1,355 @@
+#include <condition_variable>
+
+#include <chrono>
+
+#define ASSERT(expr)          GAPI_Assert(expr)
+
+#define LOG_INFO(tag, ...)    GAPI_LOG_INFO(tag, __VA_ARGS__)
+#define LOG_WARNING(tag, ...) GAPI_LOG_WARNING(tag, __VA_ARGS__)
+#define LOG_DEBUG(tag, ...)   GAPI_LOG_DEBUG(tag, __VA_ARGS__)
+
+
+#ifdef OPENCV_WITH_ITT
+const __itt_domain* cv::gimpl::parallel::gapi_itt_domain = __itt_domain_create("GAPI Context");
+#endif
+
+namespace cv{ namespace gimpl { namespace parallel {
+
+inline void assert_graph_is_running(tbb::task* root)
+{
+   //tbb::task::wait_for_all block calling thread until task ref_count is dropped to 1
+   //So if the root task ref_count is greater than 1 graph still has a job to do and
+   //according wait_for_all() has not yet returned
+   ASSERT(root->ref_count() > 1);
+}
+
+enum class use_tbb_scheduler_bypass {
+   yes,
+   no
+};
+
+enum class wake_tbb_master {
+   yes,
+   no
+};
+
+//made template to break circular dependencies
+template<typename body_t>
+struct functor_task : tbb::task {
+   body_t body;
+
+   template<typename arg_t>
+   functor_task(arg_t&& a) : body(std::forward<arg_t>(a)){}
+
+   tbb::task * execute() override
+   {
+      assert_graph_is_running(parent());
+
+      auto reuse_current_task = body();
+      //if needed, say TBB to execute current task once again
+      return (use_tbb_scheduler_bypass::yes ==  reuse_current_task) ? (recycle_as_continuation(), this) : nullptr;
+   }
+   ~functor_task(){
+      assert_graph_is_running(parent());
+   }
+};
+
+template<typename body_t>
+auto allocate_task(tbb::task* root, body_t const& body) -> functor_task<body_t>*
+{
+    return new(tbb::task::allocate_additional_child_of(*root) ) functor_task<body_t>{body};
+}
+
+template<typename body_t>
+void spawn_no_assert(tbb::task* root, body_t const& body)
+{
+   tbb::task::spawn(* allocate_task(root, body));
+}
+
+#ifdef OPENCV_WITH_ITT
+namespace {
+    static __itt_string_handle* ittTbbAddReadyBlocksToQueue       = __itt_string_handle_create( "add ready blocks to queue" );
+    static __itt_string_handle* ittTbbSpawnReadyBlocks            = __itt_string_handle_create( "spawn ready blocks" );
+    static __itt_string_handle* ittTbbEnqueueSpawnReadyBlocks     = __itt_string_handle_create( "enqueing a spawn of ready blocks" );
+    static __itt_string_handle* ittTbbUnlockMasterThread          = __itt_string_handle_create( "Unlocking master thread" );
+}
+#endif //OPENCV_WITH_ITT
+
+
+template<typename body_t>
+void batch_spawn(size_t count, tbb::task* root, body_t const& body, bool do_assert_graph_is_running = true)
+{
+   ITT_AUTO_TRACE_GUARD(ittTbbSpawnReadyBlocks);
+   if (do_assert_graph_is_running){
+       assert_graph_is_running(root);
+   }
+
+   for (size_t i=0; i<count; i++)
+   {
+       spawn_no_assert(root, body);
+   }
+}
+
+
+}}}  // namespace cv::gimpl::parallel
+
+void cv::gimpl::parallel::execute(tbb::concurrent_priority_queue<tile_node* , tile_node_indirect_priority_comparator> & q)
```

Comment:
Please reformat this code to keep line width limit. 80 is desired, ~110 is maximum.